### PR TITLE
[SPARK-35019][PYTHON][SQL]Fix type hints mismatches in pyspark.sql.*

### PR DESCRIPTION
**Pull request by Yikun**
### What changes were proposed in this pull request?
Fix type hints mismatches in pyspark.sql.*

### Why are the changes needed?
There were some mismatches in pyspark.sql.*

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
dev/lint-python passed.

### CODE REVIEW

File: python/pyspark/sql/catalog.pyi [hunk 1]
```diff
@@ -53,7 +54,7 @@ class Catalog:
     def dropGlobalTempView(self, viewName: str) -> None: ...
     def registerFunction(
         self, name: str, f: Callable[..., Any], returnType: DataType = ...
-    ) -> None: ...
+    ) -> UserDefinedFunctionLike: ...
```
Comment (Yikun) [inline python/pyspark/sql/catalog.pyi:57]: Note for reviewer:
https://github.com/apache/spark/blob/0494dc90af48ce7da0625485a4dc6917a244d580/python/pyspark/sql/catalog.py#L257

https://github.com/apache/spark/blob/ff1fc5ed4b685b4f5f83d5f600b24f089dd4522e/python/pyspark/sql/udf.pyi#L46-L51

File: python/pyspark/sql/column.pyi [hunk 2]
```diff
@@ -109,7 +109,7 @@ class Column:
         self,
         lowerBound: Union[Column, LiteralType, DateTimeLiteral, DecimalLiteral],
         upperBound: Union[Column, LiteralType, DateTimeLiteral, DecimalLiteral],
-    ) -> Column: ...
+    ) -> bool: ...
```
Comment (Yikun) [inline python/pyspark/sql/column.pyi:112]: Note for reviewer:
https://github.com/apache/spark/blob/ff1fc5ed4b685b4f5f83d5f600b24f089dd4522e/python/pyspark/sql/column.py#L801

File: python/pyspark/sql/conf.pyi [hunk 3]
```diff
@@ -21,7 +21,7 @@ from py4j.java_gateway import JavaObject  # type: ignore[import]
 
 class RuntimeConfig:
     def __init__(self, jconf: JavaObject) -> None: ...
-    def set(self, key: str, value: str) -> str: ...
+    def set(self, key: str, value: str) -> None: ...
```
Comment (Yikun) [inline python/pyspark/sql/conf.pyi:24]: https://github.com/apache/spark/blob/ff1fc5ed4b685b4f5f83d5f600b24f089dd4522e/python/pyspark/sql/conf.py#L34-L36

File: python/pyspark/sql/context.pyi [hunk 4]
```diff
@@ -63,7 +64,7 @@ class SQLContext:
     ) -> DataFrame: ...
     def registerFunction(
         self, name: str, f: Callable[..., Any], returnType: DataType = ...
-    ) -> None: ...
+    ) -> UserDefinedFunctionLike: ...
```
Comment (Yikun) [inline python/pyspark/sql/context.pyi:67]: Note for reviewer:
https://github.com/apache/spark/blob/ff1fc5ed4b685b4f5f83d5f600b24f089dd4522e/python/pyspark/sql/context.py#L236
https://github.com/apache/spark/blob/ff1fc5ed4b685b4f5f83d5f600b24f089dd4522e/python/pyspark/sql/udf.pyi#L46-L51

File: python/pyspark/sql/dataframe.pyi [hunk 5]
```diff
@@ -85,7 +85,7 @@ class DataFrame(PandasMapOpsMixin, PandasConversionMixin):
     def checkpoint(self, eager: bool = ...) -> DataFrame: ...
     def localCheckpoint(self, eager: bool = ...) -> DataFrame: ...
     def withWatermark(
-        self, eventTime: ColumnOrName, delayThreshold: str
+        self, eventTime: str, delayThreshold: str
```
Comment (Yikun) [inline python/pyspark/sql/dataframe.pyi:88]: https://github.com/apache/spark/blob/ff1fc5ed4b685b4f5f83d5f600b24f089dd4522e/python/pyspark/sql/dataframe.py#L608-L609

File: python/pyspark/sql/functions.pyi [hunk 6]
```diff
@@ -57,7 +57,7 @@ def monotonically_increasing_id() -> Column: ...
 def nanvl(col1: ColumnOrName, col2: ColumnOrName) -> Column: ...
 def percentile_approx(
     col: ColumnOrName,
-    percentage: Union[Column, float, List[float]],
+    percentage: Union[Column, float, List[float], tuple[float]],
```
Comment (Yikun) [inline python/pyspark/sql/functions.pyi:60]: Note for reviewer:
https://github.com/apache/spark/blob/ff1fc5ed4b685b4f5f83d5f600b24f089dd4522e/python/pyspark/sql/functions.py#L1218

**Review by Yikun — COMMENTED** [review]
Some note for reviwer:

File: python/pyspark/sql/dataframe.py [hunk 7]
```diff
@@ -588,7 +588,7 @@ def withWatermark(self, eventTime, delayThreshold):
 
         Parameters
         ----------
-        eventTime : str or :class:`Column`
+        eventTime : str
```
Comment (Yikun) [inline python/pyspark/sql/dataframe.py:591]: The scalar eventTime is also `str`
https://github.com/apache/spark/blob/61d038f26e1beeeb39638fdf9703d2e86d058342/sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala#L753
and python implements is also validated as string
https://github.com/apache/spark/blob/ff1fc5ed4b685b4f5f83d5f600b24f089dd4522e/python/pyspark/sql/dataframe.py#L608-L609

but wrong doc on the param and typehints

**Comment by Yikun** [comment]
cc @zero323

File: python/pyspark/sql/column.pyi (hunk 2 above)
Comment (zero323) [reply python/pyspark/sql/column.pyi:112]: This doesn't look right.

**Comment by zero323** [comment]
I've done first sweep and, with exception of `Column.between`, it looks OK.

**Comment by Yikun** [comment]
@zero323 OK, if it should be suqashed into your patch, I'd like to close this PR,  or we could just merge this first?

And looks like you have already started to fix these mismacthes for [Python type hints improvement](https://issues.apache.org/jira/browse/SPARK-34629), so I am going stop these work (fix these mismatches module by module), but if you need any help pls feel free to ping me.

**Comment by zero323** [comment]
Personally, I don't see a problemem with having this (with addressedw current and  possibly future comments) added separately.

Annotations are pretty large and not fixed thing - current state is a result of a few years of maintenance and multiple passes and there still problems that haven't been noticed so far. Not to mention there are parts that are morenmy personal an opinion about style (i.e. list vs. tuple usage or balance between false positives and false negatives), which are probably disputable if you come from different background. There are also some decisions that we made during initial port (some tests where omitted, some stubs where removed, we decided to prefer ignores instead of adding new annotations in quite a few places).

We should probably avoid fixing individual, low priority (not bug, not reported as an issue) annotations in separate PRs, to keep review work.
managable, but I don't think its a problem here. 

Also, there are cases where annotations require a lot of caution, like ML or UDFs.

I am bit preoccupied lately, but if you plan to work more on this part and have any questions, please let me know.

**Comment by Yikun** [comment]
Thanks for your detail and patient explanation for the background of type hints, learned a lot : ) and really much thanks for your previous effort on python stub works.

And I started to the annotions fix work because I saw [SPARK-34629](https://issues.apache.org/jira/browse/SPARK-34629) is `OPEN` and have a short discussion with @HyukjinKwon, and found there's some mismatches on type hints, so I thought maybe can help some on it. I believe maybe there are many other people also have will to help it.

> We should probably avoid fixing individual, low priority (not bug, not reported as an issue) annotations in separate PRs, to keep review work managable, but I don't think its a problem here.

Definately yes, I was going to fix these type annotations by module ranther than by file or smaller granularity (such as this patch, I looked in deep in every files under the pyspark.sql.*), I also considered the huge patch will hard to review, so I just submit it in the score of module. And also, I paste the origin py lines for every pyi.

> I am bit preoccupied lately, but if you plan to work more on this part and have any questions, please let me know.

I really know about the difficult of fix these (due to large amount of py files and needs to be very careful), but I'm still not sure it's the good way (I mean fix and submit it by module) to improve pyspark hints or not. If yes, I will continue this work when I have time; if not, maybe we also need to find a better way to let more people join.

File: python/pyspark/sql/column.pyi (hunk 2 above)
Comment (Yikun) [reply python/pyspark/sql/column.pyi]: Done, : )

**Comment by Yikun** [comment]
> I've done first sweep and, with exception of Column.between, it looks OK.

Addressed.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.1.

**Pull request closed without merging**